Add base entry point class for android tests

greenEkatherine opened the pull request:
Addressed #383 

With this approach test result file could be retrieved only on devices with API <=29

In scope of #385 I will add ability to pass test result file for API 30 as well

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+                    $"Ignored: {results.SkippedTests}";
+                bundle.PutString(SummaryMessageArgumentName, message);
+                
+                bundle.PutLong(ExitCodeArgumentName, results.FailedTests == 0 ? 0 : 1);
```
I guess this sort of deals with #446? Does this mean the app will exit with `1`?
And if yes, can we reference the `ExitCode` enum and use the number from there (`TestsFailed` or how is it called)?

greenEkatherine replied:
Unfortunately no, I commented in #446 how it works now. Also I replace numbers with `ExitCode` enum

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+                }
+            }
+
+            if (bundleArguments.ContainsKey(ResultsFileArgumentName) == false)
```
This `== false` seems fishy

mattleibow replied:
You could also do `bundleArguments.TryGetValue(key, out var value)` and just let the system do all the checks for you.

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            foreach (var key in arguments.KeySet())
+            {
+                string value = arguments.GetString(key);
```
Not sure where this class comes from but is there also something like `GetObject` so that we don't lose type information or it has `GetString` only?

mattleibow replied:
It has other methods https://docs.microsoft.com/en-us/dotnet/api/android.os.bundle?view=xamarin-android-sdk-9&viewFallbackFrom=xamarin-android-sdk-11

But, since this is a bundle from the OS and everything is required to be serializable, it might not change anything.

However, it might just be worth keeping a copy of the whole Bundle directly, no need to copy the values out. For the entrypoint, you could also use multitargeting to have an Android-specific variant that take a bundle and let the entrypoint worry about what it needs.

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            if (bundleArguments.ContainsKey(ResultsFileArgumentName) == false)
+            {
+                bundleArguments.Add(ResultsFileArgumentName, "TestResults.xml");
```
I think we use `testResults.xml` in other places. Can you please check for the `T` or `t`?

mattleibow replied:
I know the runtime may be different, but the standard for xunit tests is `T`. I think. At least it was that with NUnit.

greenEkatherine replied:
I've checked in XHarness we use TestResults for NUnit and XUnit, so I leave it as it is

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+                bundle.PutString(ResultsPathArgumentName, entryPoint.TestsResultsFinalPath);
+            }
+            
+            if (bundle.GetLong(ExitCodeArgumentName, -1) == -1)
```
I think we had `ExitCode` for `ExitCodeNotSet`. Is this `1` expected by the Android API or it's up to us?

greenEkatherine replied:
I've replaced it with `ExitCodeNotSet`, it's up to us I guess

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            protected override TestRunner GetTestRunner(LogWriter logWriter)
+            {
+                var testRunner = base.GetTestRunner(logWriter);
```
Please compare with the `iOSApplicationEntryPoint`. When we were ignoring some methods, we would also flip some flag saying "do not execute all tests". Not sure we should do the same here

greenEkatherine replied:
I added handling with `runAllTests`

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+            protected override IEnumerable<TestAssemblyInfo> GetTestAssemblies()
+            {
+                //yield return new TestAssemblyInfo(Assembly.GetExecutingAssembly(), Assembly.GetExecutingAssembly().Location);
+                //yield return new TestAssemblyInfo(typeof(Battery_Tests).Assembly, typeof(Battery_Tests).Assembly.Location);
```
Forgotten code?

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+        class TestsEntryPoint : AndroidApplicationEntryPoint
+        {
+            readonly string resultsPath;
```
nit: I think this project requires visibility keywords explicitly and private fields to begin with `_`. The settings in `.editorconfig` and VS should hint you

mandel-macaque replied:
While private is the default, we indeed use the explicit form due to the netcore style.

greenEkatherine replied:
Now I get it why it was underlined in VS 😃

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+                    Directory.CreateDirectory(docsDir);
+                }
+
+                resultsPath = Path.Combine(docsDir, arguments[ResultsFileArgumentName]);
```
Does this mean all of the different bundles will keep overwriting the same testResults file?

greenEkatherine replied:
Not necessary because we just provide default name and for each test different names could be provided from the parameter. But I may try to generate somehow the name from test info

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit.csproj:
```diff
@@ -0,0 +1,20 @@
+
+  <ItemGroup>
+    <Reference Include="Mono.Android">
+      <HintPath>..\..\..\..\..\..\..\Program Files (x86)\Microsoft Visual Studio\2019\Preview\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v11.0\Mono.Android.dll</HintPath>
```
As discussed on the sync, this looks fishy and won't work for non-Windows platforms. Can we get this from a NuGet maybe?

greenEkatherine replied:
It seems that there is no suitable NuGet for this. Since I use `Instrumentation` and `Bundle` from there I should rework the change significantly. Let me discuss this with @mattleibow

premun replied:
Yeah, there must be a way since Matthew also must be implementing it somehow in the place that you are trying to now replace

mattleibow replied:
Think think it is because I am building an Android library. Basically using msbuild.sdk.extras with a tfm of monoandroid10.0

mattleibow replied:
You should be able to multi target across netstd and monoandroid.

greenEkatherine replied:
Removed this dependency and targeted monoandroid

greenEkatherine replied:
It is not preserved after update

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit.csproj:
```diff
@@ -0,0 +1,20 @@
+
+  <PropertyGroup>
+    <IsPackable>true</IsPackable>
+    <TargetFrameworks>netstandard2.1;netcoreapp3.1</TargetFrameworks>
```
If the project stays, it should probably be only `netstandard2.1`:
```suggestion
    <TargetFrameworks>netstandard2.1</TargetFrameworks>
```

mattleibow replied:
Not sure if we can use msbuild.sdk.extras?

mattleibow replied:
might be worth just doing `netstandard2.1;monoandroid10.0;xamarin.ios10`. Then packing it will make sure to copy all the files to the right place and each OS gets what it needs.

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+        protected Dictionary<string, string> bundleArguments;
+
+        protected AndroidInstrumentationBase()
```
Do we need/want this ctor?

premun reviewed (commented):
Couple of questions left. Also please consider describing the usage in the README

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            public TestsEntryPoint(Dictionary<string, string> arguments)
+            {
+                var root = Application.Context.GetExternalFilesDir(null)?.AbsolutePath;
```
? implies that root is nullable, right? Looks like it is not considered later in the file, for example, Path.Combine will throw a NRE parameter name path1.

premun replied:
This is a good catch. @greenEkatherine can you please turn on the nullability for this project?

It is done by adding `<nullable>enable</nullable>` in the csproj (please check the other csproj, I'm not sure it's `enable` or `enabled`)

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            protected override bool LogExcludedTests => true;
+
+            public override TextWriter Logger => null;
```
Hmm do we have #enable nullable? If that is the cqse TextWriter cannot be null, if it is, we will probably have issues. If indeed can be null, we should add TextWriter?

We should use nullable references when possible: https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            public override string TestsResultsFinalPath => resultsPath;
+
+            protected override int? MaxParallelThreads => System.Environment.ProcessorCount;
```
Careful! What happens if we add to many threads? The OS (in this case Android) will decide if an application is too greedy, if it considers that, it will terminate the application. That would result in a bad test run, we prefer slower test runs that will always finish. 

This has happened on iOS, will probably happen on Android too.

mattleibow replied:
Is it safe to do something like `max/2 + 1`?

greenEkatherine replied:
Changed to half of the threads

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+
+            protected override int? MaxParallelThreads => System.Environment.ProcessorCount;
+
+            protected override IDevice Device => null;
```
Same, nullable references, we should make sure that this will not result in a NRE, if it does not, lets add IDevice?

greenEkatherine replied:
I checked, it could be null and don't be cause of exceptions

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+            protected override TestRunner GetTestRunner(LogWriter logWriter)
+            {
+                var testRunner = base.GetTestRunner(logWriter);
+                if (excludeCategories != null)
+                {
+                    testRunner.SkipCategories(excludeCategories.Split(' '));
+                }
+                if (excludeMethod != null)
+                {
+                    foreach (var method in excludeMethod.Split(' '))
+                    {
+                        testRunner.SkipMethod(method, true);
+                    }
+                }
+                if (excludeClass != null)
+                {
+                    foreach (var item in excludeClass.Split(' '))
+                    {
+                        testRunner.SkipMethod(item, true);
+                    }
+                }
+                if (includeMethod != null)
+                {
+                    foreach (var method in includeMethod.Split(' '))
+                    {
+                        testRunner.SkipMethod(method, false);
+                    }
+                }
+                if (includeClass != null)
+                {
+                    foreach (var item in includeClass.Split(' '))
+                    {
+                        testRunner.SkipMethod(item, false);
+                    }
+                }
+                return testRunner;
+            }
```
Just as an exercise, you do not need to write that many times the same code, you can see the pattern:
```csharp
if (filter != null) {
   foreach (var f in fiter.Split (' ') {
      Lambda (f, included);
   }
}
```
where Lambda:
```csharp
Lambda = Action<string, bool> ()
```
For ref: https://docs.microsoft.com/en-us/dotnet/api/system.action-2?view=net-5.0

So, you can get all those lines in a single method, and no need to copy paste which is error prone.

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+                    $"Passed: {results.PassedTests} " +
+                    $"Inconclusive: {results.InconclusiveTests} " +
+                    $"Failed: {results.FailedTests} " +
+                    $"Ignored: {results.SkippedTests}";
```
Exercise, how many times and how many strings are we using here? 

$"Text {var}" does something you are not expecting: https://sharplab.io/#gist:94115c1be673ee9bcc131ce02d9dc433

I count 5 calls to format and 1 call to contact while you can do 1 call to Format. Is a nit, but is good to use PRs to learn the platform.

mattleibow commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,198 @@
+        {
+        }
+
+        protected virtual IEnumerable<string> Tests
```
Maybe this should be Assembly instead? Do all assemblies have a physical location? What about thinks like AOT and is there a case where the assemblies are read from a stream out of the apk? Maybe @jonathanpeppers  or @jonpryor  has a comment? Or even just use `TestAssemblyInfo` directly.

jonathanpeppers replied:
In some cases, this path would not be useful. `.dll` files remain in the `.apk` on disk and you can't pull them with `adb pull` or anything.

In what way is this path used?

greenEkatherine replied:
Agree, replaced with Assembly

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,192 @@
+                    return new List<Assembly>();
+                }
+                internal set { }
+            }
```
This property looks very very strange. Lets thing about it for a second. 

get => Will always return a new List, if we for example do the following:

```csharp
var first = this.Tests;
var second = this.Tests;
Object.ReferenceEquals (first, first); // will return true
Object.ReferenceEquals (first, second); // will return false
```
Therefore we do have 2 instances that are different, either we do not want a property, we have an unusual behaviour or we got a bug.

Lets know look at the set. The way you wrote it is doing **nothing**. So 
```csharp
this.Test = new List<Assembly>(){"foo", "bar"}; // will have no effect, we alway return an empty list
var first = this.Test; // results in an empty list
```
You probably wanted to do the following:
```csharp
public IEnumerable<Assembly> Tests { get; set; } = new List<Assembly>();
```
The **get** and the **set** will be generated by the compiler, and it will be init to be an empty list. Take a look at https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties

greenEkatherine replied:
fixed

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,192 @@
+        {
+        }
+
+        protected virtual IEnumerable<Assembly> Tests
+        {
+            get
+            {
+                yield return Assembly.GetExecutingAssembly();
+            }
```
I'm confuse by this property. Just returns the currently executing assembly. What is the goal? I'm not an android expert. But either:

1. We have on instrumentation with several test assemblies (each one of them with several tests).
2. We have an instrumentation class per assembly, with a number of tests.

If the second is correct, we could  modify the property to be just Assembly, else.. I don't understand why the class returns a IEnumerable.

greenEkatherine replied:
This property was aimed to pass tests from user to `TestsEntryPoint::GetTestAssemblies`. Anyway I'm going to remove `Instrumentation` class from this pull request to adjust it with #480

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,192 @@
+
+        class TestsEntryPoint : AndroidApplicationEntryPoint
+        {
+            readonly string _resultsPath;
+            readonly string _excludeCategoriesDir;
+            readonly string _excludeCategoriesFile;
```
I think @premun  whats a 'private' in those:
```suggestion
            private readonly string _resultsPath;
            private readonly string _excludeCategoriesDir;
            private readonly string _excludeCategoriesFile;
```

mandel-macaque commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs:
```diff
@@ -0,0 +1,192 @@
+                }
+            }
+
+            protected override TestRunner GetTestRunner(LogWriter logWriter)
+            {
+                var testRunner = base.GetTestRunner(logWriter);
+
+                ConfigureFilters(_parsedArguments[ExcludeMethodArgumentName], testRunner.SkipMethod, true);
+                ConfigureFilters(_parsedArguments[ExcludeClassArgumentName], testRunner.SkipClass, true);
+                ConfigureFilters(_parsedArguments[IncludeMethodArgumentName], testRunner.SkipMethod, false);
+                ConfigureFilters(_parsedArguments[IncludeClassArgumentName], testRunner.SkipClass, false);
+
+                return testRunner;
+            }
```
LOVE IT!

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/DefaultAndroidEntryPoint.cs:
```diff
@@ -0,0 +1,98 @@
+using Microsoft.DotNet.XHarness.TestRunners.Common;
+using Microsoft.DotNet.XHarness.TestRunners.Xunit;
+
+namespace Microsoft.DotNet.XHarness.DefaultAndroidEntryPoint.Xunit
```
Can you please add a documentation header explaining the purpose of this class?
I think it will be very important for anyone looking into this.

Also please make a note in the README, where we talk about TestRunners. Otherwise people who are not Matthew won't know about this

greenEkatherine replied:
added

premun commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/DefaultAndroidEntryPoint.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+            var testRunner = base.GetTestRunner(logWriter);
+
+            Tuple<string, Action<string, bool>, bool>[] filters =
```
FYI, in C# 8.0+, you can use tuples without the Tuple class, you can even name the items.

The initialization would be:
```csharp
(string, Action<string, bool>, bool)[] filters
```
or with named version:
```csharp
(string Filter, Action<string, bool> FilterMethod, bool IsExcluded)[] filters
```
And the initialization would then be also shorter:
```csharp
(_parsedArguments.GetValueOrDefault(ExcludeMethodArgumentName), testRunner.SkipMethod, true),
```
Below, you would then do:
```csharp
ConfigureFilters(t.Filter, t.FilterMethod, t.IsExcluded);
```

premun replied:
Or, please also consider creating a private class/record inside of this class:
```csharp
private class FilterDefinition { ...
```

And then use it here and in the ConfigureFilters method.

premun replied:
Although, this ConfigureFilters is so short it could be inlined as well without problems too.

greenEkatherine replied:
Cool, I was looking for something like this initialization, thanks!

greenEkatherine replied:
done

premun reviewed (approved):
Looks clean!

mattleibow commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/DefaultAndroidEntryPoint.cs:
```diff
@@ -0,0 +1,116 @@
+                Directory.CreateDirectory(_parsedArguments[ResultsFileArgumentPath]);
+            }
+
+            _resultsPath = Path.Combine(_parsedArguments[ResultsFileArgumentPath], _parsedArguments[ResultsFileArgumentName]);
```
There is no checks for this, so it will just crash if not provided. I'm thinking this should just be a string file path on the ctor.

greenEkatherine replied:
replaced

mattleibow reviewed (commented):
Quite good. Just a comment on the visibility of the args.

But I am also not sure how the filter from the CLI gets to the entry point. I added a `--method` and it never exists on any of the command lines generated. If the CLI runner is going to pass them, then maybe we need to coordinate with what the CLI will use?

greenEkatherine commented:
> But I am also not sure how the filter from the CLI gets to the entry point. I added a `--method` and it never exists on any of the command lines generated. If the CLI runner is going to pass them, then maybe we need to coordinate with what the CLI will use?

`--class` and `--method` parameters are filtering for apple commands only. I'm sorry for the confusion, there are some mix of arguments in `help` command now but it's going to be fixed in #431 

However, `TestRunner` has more options to work with classes and methods filtering, so I added the possibility to use it in that way.

greenEkatherine commented on src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/DefaultAndroidEntryPoint.cs:
```diff
@@ -0,0 +1,116 @@
+        private const string ExcludeMethodArgumentName = "exclude-method";
+        private const string IncludeMethodArgumentName = "include-method";
+        private const string ExcludeClassArgumentName = "exclude-class";
+        private const string IncludeClassArgumentName = "include-class";
```
agree, done

mattleibow commented on XHarness.sln:
```diff
@@ -42,6 +42,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DotNet.XHarness.A
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DotNet.XHarness.DefaultAndroidEntryPoint.Xunit", "src\Microsoft.DotNet.XHarness.InstrumentationBase.Xunit\Microsoft.DotNet.XHarness.DefaultAndroidEntryPoint.Xunit.csproj", "{FBDF39C0-1096-4088-B057-2FF8CCDFB32D}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Xamarin.Essentials.DeviceTests.Android", "..\..\..\Downloads\APK\Xamarin.Essentials.DeviceTests.Android.csproj", "{CB2072E0-A437-4811-AE17-16CAE0DDA1B1}"
+EndProject
```
I think this was not supposed to be here?

greenEkatherine replied:
yep, removing

mattleibow commented:
Very nice! Thanks for all of this!
